File: tmux_eyes/backend.py
"""Multiplexer backend selection.

Factory that returns a :class:`MultiplexerClientProto` instance based on
config + environment, so the rest of the daemon doesn't care whether the
underlying terminal multiplexer is tmux or wezterm.
"""

from __future__ import annotations

import logging
import os
import shutil
from typing import TYPE_CHECKING

from .config import Config
from .types import MultiplexerClientProto

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)


class BackendError(RuntimeError):
    """No usable multiplexer backend could be selected."""


def detect_backend() -> str:
    """Best-effort autodetection: returns 'tmux' or 'wezterm'.

    Order of precedence:
      1. ``$TMUX`` env var set → tmux (we're inside a tmux session right now).
      2. ``$WEZTERM_PANE`` set → wezterm (we're inside a wezterm pane).
      3. ``tmux`` binary on PATH → tmux.
      4. ``wezterm`` binary on PATH → wezterm.
      5. Raise BackendError.
    """
    if os.environ.get("TMUX"):
        return "tmux"
    if os.environ.get("WEZTERM_PANE"):
        return "wezterm"
    if shutil.which("tmux"):
        return "tmux"
    if shutil.which("wezterm"):
        return "wezterm"
    raise BackendError(
        "No multiplexer detected. Set TMUX_EYES_BACKEND=tmux or =wezterm, "
        "or install one of them."
    )


def select_backend(config: Config) -> MultiplexerClientProto:
    """Build a multiplexer client based on the resolved backend name."""
    name = (config.backend or "auto").lower()
    if name == "auto":
        name = detect_backend()
        logger.info("backend autodetected: %s", name)

    if name == "tmux":
        from .tmux_io import TmuxClient

        return TmuxClient(socket=config.tmux_socket)

    if name == "wezterm":
        from .wezterm_io import WeztermClient

        return WeztermClient()

    raise BackendError(
        f"Unknown backend {name!r}. Use 'tmux', 'wezterm', or 'auto'."
    )

File: rs/src/backend.rs
//! Multiplexer backend selection.
//!
//! Factory that returns a `Box<dyn MultiplexerClient>` based on the resolved
//! backend name, so the rest of the daemon doesn't care whether the underlying
//! terminal multiplexer is tmux or wezterm.

use crate::config::Config;
use crate::multiplexer::tmux::TmuxClient;
use crate::multiplexer::wezterm::WeztermClient;
use crate::types::MultiplexerClient;

// ---------------------------------------------------------------------------
// PATH search helper (no `which` crate available)
// ---------------------------------------------------------------------------

fn binary_in_path(name: &str) -> bool {
    if let Ok(path) = std::env::var("PATH") {
        for dir in std::env::split_paths(&path) {
            if dir.join(name).is_file() {
                return true;
            }
            #[cfg(windows)]
            if dir.join(format!("{}.exe", name)).is_file() {
                return true;
            }
        }
    }
    false
}

// ---------------------------------------------------------------------------
// Detection
// ---------------------------------------------------------------------------

/// Detect the best available multiplexer backend.
///
/// Order of precedence:
///   1. `$TMUX` env var set → tmux (we're inside a tmux session).
///   2. `$WEZTERM_PANE` set → wezterm (we're inside a wezterm pane).
///   3. `tmux` binary on PATH → tmux.
///   4. `wezterm` binary on PATH → wezterm.
///   5. Error.
pub fn detect_backend() -> anyhow::Result<&'static str> {
    if std::env::var("TMUX").is_ok() {
        return Ok("tmux");
    }
    if std::env::var("WEZTERM_PANE").is_ok() {
        return Ok("wezterm");
    }
    if binary_in_path("tmux") {
        return Ok("tmux");
    }
    if binary_in_path("wezterm") {
        return Ok("wezterm");
    }
    anyhow::bail!(
        "no multiplexer detected — set TMUX_EYES_BACKEND=tmux or wezterm"
    )
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/// Build a `MultiplexerClient` based on `cfg.backend`.
///
/// `"auto"` triggers [`detect_backend`]. `"tmux"` and `"wezterm"` are explicit.
pub fn select_backend(cfg: &Config) -> anyhow::Result<Box<dyn MultiplexerClient>> {
    let name = if cfg.backend.is_empty() {
        "auto".to_string()
    } else {
        cfg.backend.to_lowercase()
    };

    let resolved = if name == "auto" {
        detect_backend()?.to_string()
    } else {
        name
    };

    match resolved.as_str() {
        "tmux" => {
            let client = TmuxClient::new(&cfg.tmux_socket)?;
            Ok(Box::new(client))
        }
        "wezterm" => Ok(Box::new(WeztermClient::new())),
        other => anyhow::bail!(
            "Unknown backend {:?}. Use 'tmux', 'wezterm', or 'auto'.",
            other
        ),
    }
}
